### PR TITLE
Distro Module

### PR DESCRIPTION
**Pull request by coachchucksol**

### CODE REVIEW

File: program/src/distribute_ncn_vault_rewards.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    {
+        msg!("rewards {}", rewards);
+        msg!("vault {}", vault.key);
+        msg!("vault {}", vault.lamports.borrow_mut());
+        msg!("router {}", ncn_reward_router.key);
+        msg!("router {}", ncn_reward_router.lamports.borrow_mut());
+        msg!(
+            "rent {}",
```
Comment (ebatsell) [inline program/src/distribute_ncn_vault_rewards.rs:79]: remove this and commented out code below

File: program/src/register_mint.rs [hunk 2]
```diff
@@ -85,6 +85,8 @@ pub fn process_register_mint(program_id: &Pubkey, accounts: &[AccountInfo]) -> P
     let vault_data = vault.data.borrow();
     let vault = Vault::try_from_slice_unchecked(&vault_data)?;
 
+    //TODO add logic to sort into group
```
Comment (ebatsell) [inline program/src/register_mint.rs:88]: for future PR?

File: program/src/set_config_fees.rs [hunk 3]
```diff
@@ -53,11 +59,25 @@ pub fn process_set_config_fees(
...
     }
 
-    config.fees.set_new_fees(
-        new_dao_fee_bps,
-        new_ncn_fee_bps,
+    let base_fee_group = if let Some(base_fee_group) = base_fee_group {
+        Some(BaseFeeGroup::try_from(base_fee_group)?)
+    } else {
+        None
+    };
```
Comment (ebatsell) [inline program/src/set_config_fees.rs:66]: nit: .map() could be a bit cleaner here

File: program/src/distribute_base_ncn_reward_route.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    account_info::AccountInfo, clock::Clock, entrypoint::ProgramResult, msg,
+    program_error::ProgramError, pubkey::Pubkey, sysvar::Sysvar,
+};
+
+/// Can be backfilled for previous epochs
+pub fn process_distribute_base_ncn_reward_route(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    ncn_fee_group: u8,
+    first_slot_of_ncn_epoch: Option<u64>,
```
Comment (ebatsell) [inline program/src/distribute_base_ncn_reward_route.rs:17]: IMO makes more sense just pass in the epoch itself here as a non optional value

**Review by ebatsell — COMMENTED** [review]
Remove test.txt

File: core/src/ballot_box.rs [hunk 5]
```diff
@@ -463,17 +479,23 @@ impl BallotBox {
...
-    pub fn has_ballot(&self, ballot: &Ballot) -> bool {
-        self.ballot_tallies.iter().any(|t| t.ballot.eq(ballot))
+        // // Check that the merkle root is one of the existing options
+        // // TODO do we want to be able to set a tie breaker that wasn't voted on?
+        // // What if there were malicious merkleroots?
+        // if !self.has_ballot(&finalized_ballot) {
+        //     return Err(TipRouterError::TieBreakerNotInPriorVotes);
+        // }
+
+        self.set_winning_ballot_tally(*finalized_ballot_tally);
```
Comment (ebatsell) [inline core/src/ballot_box.rs:497]: what's the reasoning for changing to use this method?

File: core/src/ncn_reward_router.rs [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+            .to_imprecise()
+            .ok_or(TipRouterError::CastToImpreciseNumberError)?;
+        let dao_rewards: u64 = dao_rewards_u128
+            .try_into()
+            .map_err(|_| TipRouterError::CastToU64Error)?;
+
+        Ok(dao_rewards)
+    }
+
+    fn calculate_vault_reward(
```
Comment (ebatsell) [inline core/src/ncn_reward_router.rs:261]: function names for this and calculate_operator_rewards seem off based on what they're returning

File: program/src/distribute_ncn_operator_rewards.rs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        program_id,
+        ncn_fee_group,
+        operator.key,
+        ncn.key,
+        ncn_epoch,
+        ncn_reward_router,
+        true,
+    )?;
+
+    //TODO do we want an Operator Fee Wallet?
```
Comment (ebatsell) [inline program/src/distribute_ncn_operator_rewards.rs:49]: TODO

File: program/src/distribute_ncn_vault_rewards.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+    if restaking_program.key.ne(&jito_restaking_program::id()) {
+        msg!("Incorrect restaking program ID");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    if vault_program.key.ne(&jito_vault_program::id()) {
+        msg!("Incorrect vault program ID");
+        return Err(ProgramError::InvalidAccountData);
+    }
```
Comment (ebatsell) [inline program/src/distribute_ncn_vault_rewards.rs:34]: curious if there's purpose to passing in the vault program as account if we're gonna require it to be a certain value and don't need to invoke the program?

File: program/src/initialize_ncn_reward_router.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    let has_winning_ballot = {
+        let ballot_box_data = ballot_box.data.borrow();
+        let ballot_box = BallotBox::try_from_slice_unchecked(&ballot_box_data)?;
+        ballot_box.has_winning_ballot()
+    };
+
+    if !has_winning_ballot {
+        msg!("Ballot has to be finalized before initializing ncn reward router");
+        return Err(ProgramError::InvalidAccountData);
+    }
```
Comment (ebatsell) [inline program/src/initialize_ncn_reward_router.rs:58]: could we get rid of this check?

File: core/src/instruction.rs [hunk 10]
```diff
@@ -23,22 +23,32 @@ pub enum TipRouterInstruction {
...
-        new_dao_fee_bps: Option<u64>,
-        new_ncn_fee_bps: Option<u64>,
-        new_block_engine_fee_bps: Option<u64>,
-        new_fee_wallet: Option<Pubkey>,
+    new_block_engine_fee_bps: Option<u16>,
+    base_fee_group: Option<u8>,
+    new_base_fee_wallet: Option<Pubkey>,
+    new_base_fee_bps: Option<u16>,
+    ncn_fee_group: Option<u8>,
+    new_ncn_fee_bps: Option<u16>,
```
Comment (ebatsell) [inline core/src/instruction.rs:51]: formatting

File: program/src/distribute_ncn_vault_rewards.rs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        program_id,
+        ncn_fee_group,
+        operator.key,
+        ncn.key,
+        ncn_epoch,
+        ncn_reward_router,
+        true,
+    )?;
+
+    //TODO do we want an Operator Fee Wallet?
```
Comment (ebatsell) [inline program/src/distribute_ncn_vault_rewards.rs:56]: TODO

File: program/src/set_config_fees.rs (hunk 3 above)
Comment (coachchucksol) [reply program/src/set_config_fees.rs:66]: done

File: program/src/distribute_base_ncn_reward_route.rs (hunk 4 above)
Comment (coachchucksol) [reply program/src/distribute_base_ncn_reward_route.rs:17]: Will change in refactor

File: core/src/ballot_box.rs (hunk 5 above)
Comment (coachchucksol) [reply core/src/ballot_box.rs]: Changed back

File: core/src/ncn_reward_router.rs (hunk 6 above)
Comment (coachchucksol) [reply core/src/ncn_reward_router.rs:285]: updated names

File: program/src/distribute_ncn_operator_rewards.rs (hunk 7 above)
Comment (coachchucksol) [reply program/src/distribute_ncn_operator_rewards.rs:49]: Done

File: program/src/distribute_ncn_vault_rewards.rs (hunk 8 above)
Comment (coachchucksol) [reply program/src/distribute_ncn_vault_rewards.rs:34]: vault_program is needed to call `Vault::load`

File: program/src/initialize_ncn_reward_router.rs (hunk 9 above)
Comment (coachchucksol) [reply program/src/initialize_ncn_reward_router.rs:58]: We can, however, closing this account will need the ballot box

Comment (coachchucksol) [reply program/src/initialize_ncn_reward_router.rs:58]: taking out for now

File: program/src/distribute_ncn_vault_rewards.rs (hunk 11 above)
Comment (coachchucksol) [reply program/src/distribute_ncn_vault_rewards.rs:56]: Done

File: core/src/instruction.rs (hunk 10 above)
Comment (coachchucksol) [reply core/src/instruction.rs:51]: done

File: program/src/distribute_ncn_vault_rewards.rs (hunk 8 above)
Comment (ebatsell) [reply program/src/distribute_ncn_vault_rewards.rs:34]: sure but if the instruction fails if we pass in anything other than vault::id() why not just pass vault::id() into Vault::load and save an account from the instruction? (same question for the restaking program checks, but prob getting rid of those when we factor out ncn_epoch)

**Comment by ebatsell** [comment]
also if you're able to update the package complained about in the security audit please do, if it's not possible then we can add it to the list of ignores

File: program/src/distribute_ncn_vault_rewards.rs (hunk 8 above)
Comment (coachchucksol) [reply program/src/distribute_ncn_vault_rewards.rs:34]: good point

File: program/src/route_base_rewards.rs [hunk 12]
```diff
@@ -50,10 +50,13 @@ pub fn process_route_base_rewards(
 
     let rent_cost = base_reward_router_account.rent_cost(&Rent::get()?)?;
 
+    msg!("incoming");
     base_reward_router_account.route_incoming_rewards(rent_cost, base_reward_router_balance)?;
 
+    msg!("reward pool");
     base_reward_router_account.route_reward_pool(epoch_snapshot_account.fees())?;
 
+    msg!("ncn fee group rewards");
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/route_base_rewards.rs:59]: done

File: program/src/distribute_base_ncn_reward_route.rs [hunk 13]
```diff
@@ -57,13 +57,14 @@ pub fn process_distribute_base_ncn_reward_route(
...
 
-    if rewards == 0 {
-        msg!("No rewards to distribute");
-        return Err(TipRouterError::NoRewards.into());
-    }
+    //TODO should this be an error?
+    // if rewards == 0 {
+    //     msg!("No rewards to distribute");
+    //     return Err(TipRouterError::NoRewards.into());
+    // }
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply program/src/distribute_base_ncn_reward_route.rs:64]: I think so

File: core/src/base_reward_router.rs [hunk 14]
```diff
@@ -202,6 +217,21 @@ impl BaseRewardRouter {
...
+                    //     "Route NCN Fee Group Rewards: {:?} {:?} {:?} {}/{} {}/{}",
+                    //     group,
+                    //     operator,
+                    //     ncn_fee_group_route_reward,
+                    //     ncn_route_reward_stake_weight,
+                    //     winning_reward_stake_weight,
+                    //     self.ncn_fee_group_rewards(*group)?,
+                    //     rewards_to_process
+                    // );
+                    // msg!("   ");
```
[the remark this answers is not in the document]
Comment (coachchucksol) [reply core/src/base_reward_router.rs:233]: Done

**Review by ebatsell — APPROVED** [review]

**Pull request merged**